File: tools/teams/weapon_rules.py
"""
Weapon-rule token parser.

Turns the verbatim `WR` cell of a Wahapedia datacard weapon row into
`WeaponRule[]` as declared in src/core/types.ts:

    { id: WeaponRuleId, x?: number, dist?: number, only?: string, raw: string }

The 23 universal rules are the Appendix › WEAPON RULES list
(docs/context-pack/research/core-rules/appendix.txt), transcribed verbatim in
UNIVERSAL below. `Piercing Crits x` and `Seek Light` are the two declared
sub-forms and map to the distinct ids `PiercingCrits` / `SeekLight`.

Anything else is a RARE rule: it keeps a stable PascalCase id and is emitted to
data/teams/_rare-weapon-rules.json together with the team(s) that use it and the
verbatim definition text. Nothing is ever dropped and nothing becomes `_unknown`.
"""

from __future__ import annotations

import re

# id -> takes a numeric parameter?
UNIVERSAL: dict[str, bool] = {
    "Accurate": True,
    "Balanced": False,
    "Blast": True,      # inches
    "Brutal": False,
    "Ceaseless": False,
    "Devastating": True,
    "Heavy": False,     # optional "(x only)" restriction
    "Hot": False,
    "Lethal": True,     # "5+"
    "Limited": True,
    "Piercing": True,
    "PiercingCrits": True,
    "Punishing": False,
    "Range": True,      # inches
    "Relentless": False,
    "Rending": False,
    "Saturate": False,
    "Seek": False,
    "SeekLight": False,
    "Severe": False,
    "Shock": False,
    "Silent": False,
    "Stun": False,
    "Torrent": True,    # inches
}

# printed spelling -> id (longest first at match time)
_NAME_TO_ID = {
    "piercing crits": "PiercingCrits",
    "seek light": "SeekLight",
    **{k.lower(): k for k in UNIVERSAL if k not in ("PiercingCrits", "SeekLight")},
}

_SPLIT = re.compile(r",(?![^()]*\))")
_MARKER = re.compile(r"(\*+|\^\d+(?:,\d+)*)\s*$")
_DIST_PREFIX = re.compile(r'^(\d+(?:\.\d+)?)\s*"\s+')
_ONLY = re.compile(r"\(([^)]*?)\s+only\)\s*$", re.I)
_NUM = re.compile(r'^(.*?)\s+(\d+(?:\.\d+)?)\s*(?:"|\+)?$')


def rare_id(name: str) -> str:
    """'Concealed Position' -> 'ConcealedPosition'; 'Anti-PSYKER' -> 'AntiPSYKER'."""
    parts = re.split(r"[^0-9A-Za-z]+", name)
    out = []
    for p in parts:
        if not p:
            continue
        out.append(p if p.isupper() or p[0].isupper() else p[0].upper() + p[1:])
    return "".join(out) or "Unnamed"


def parse_token(tok: str) -> dict | None:
    """One comma-separated WR token -> a rule dict (plus scraper-only extras
    `rare` and `marker`, which normalise.py strips before writing)."""
    raw = tok.strip()
    if not raw or raw == "-":
        return None
    body = raw

    marker = ""
    m = _MARKER.search(body)
    if m:
        marker = m.group(1)
        body = body[: m.start()].strip()

    dist = None
    m = _DIST_PREFIX.match(body)
    if m:
        dist = float(m.group(1))
        body = body[m.end():].strip()

    only = None
    m = _ONLY.search(body)
    if m:
        only = m.group(1).strip()
        # canonicalise the printed casing ("Dash Only" vs "Dash only")
        only = only[0].upper() + only[1:] if only else only
        body = body[: m.start()].strip()

    x = None
    m = _NUM.match(body)
    if m and m.group(1).strip():
        name, num = m.group(1).strip(), m.group(2)
        if name.lower() in _NAME_TO_ID:
            body, x = name, float(num) if "." in num else int(num)
    lookup = body.lower()

    if lookup in _NAME_TO_ID:
        rid = _NAME_TO_ID[lookup]
        rule: dict = {"id": rid, "raw": raw}
        if x is not None:
            rule["x"] = x
        if dist is not None:
            rule["dist"] = dist
        if only is not None:
            rule["only"] = only
        rule["rare"] = False
        rule["marker"] = marker
        rule["name"] = rid
        return rule

    rule = {"id": rare_id(body), "raw": raw}
    if x is not None:
        rule["x"] = x
    if dist is not None:
        rule["dist"] = dist
    if only is not None:
        rule["only"] = only
    rule["rare"] = True
    rule["marker"] = marker
    rule["name"] = body
    return rule


def parse_wr(wr: str) -> list[dict]:
    if not wr or wr.strip() in ("-", ""):
        return []
    out = []
    for tok in _SPLIT.split(wr):
        r = parse_token(tok)
        if r:
            out.append(r)
    return out


def public_rule(rule: dict) -> dict:
    """Drop the scraper-only keys so the JSON matches `WeaponRule` exactly."""
    out = {"id": rule["id"]}
    for k in ("x", "dist", "only"):
        if k in rule:
            out[k] = rule[k]
    out["raw"] = rule["raw"]
    return out

File: src/core/weaponRules.ts
/**
 * Weapon-rule parsing + registry.
 *
 * Appendix › WEAPON RULES lists 23 universal rules. Anything else on a datacard is a RARE
 * rule belonging to a kill team; it must be registered by that team's module. An unknown
 * token is a data-lint failure (`assertKnownRules`), never a silent no-op.
 */
import type { WeaponRule, WeaponRuleId } from './types.ts';

export const UNIVERSAL_WEAPON_RULES = [
  'Accurate',
  'Balanced',
  'Blast',
  'Brutal',
  'Ceaseless',
  'Devastating',
  'Heavy',
  'Hot',
  'Lethal',
  'Limited',
  'Piercing',
  'PiercingCrits',
  'Punishing',
  'Range',
  'Relentless',
  'Rending',
  'Saturate',
  'Seek',
  'SeekLight',
  'Severe',
  'Shock',
  'Silent',
  'Stun',
  'Torrent',
] as const satisfies readonly WeaponRuleId[];

const UNIVERSAL_SET = new Set<string>(UNIVERSAL_WEAPON_RULES);

/**
 * Rare rules registered by team modules, with the verbatim text for the tooltip.
 *
 * The SAME rule id can be printed differently by different kill teams, so the text is kept per
 * team as well as generically. Anti-PSYKER is the proven case: the Novitiates' Condemnor prints
 * "…add 1 to both Dmg stats of this weapon and it has the Lethal 5+ weapon rule", while the
 * Celestian Insidiants' own footnote prints only "…it has the Lethal 5+ weapon rule"
 * (docs/DECISIONS.md D-025). Showing one team the other's wording is simply wrong.
 */
const rareRegistry = new Map<string, { text: string; teams: Set<string>; byTeam: Map<string, string> }>();

export function registerRareWeaponRule(id: string, text: string, teamId: string): void {
  const entry = rareRegistry.get(id) ?? { text, teams: new Set<string>(), byTeam: new Map<string, string>() };
  entry.text ||= text;
  entry.teams.add(teamId);
  if (text) entry.byTeam.set(teamId, text);
  rareRegistry.set(id, entry);
}

/** The verbatim definition as THIS kill team prints it, falling back to the generic one. */
export function rareWeaponRuleText(id: string, teamId?: string): string | undefined {
  const entry = rareRegistry.get(id);
  if (!entry) return undefined;
  return (teamId ? entry.byTeam.get(teamId) : undefined) ?? entry.text;
}

export function isKnownWeaponRule(id: string): boolean {
  return UNIVERSAL_SET.has(id) || rareRegistry.has(id);
}

export function rareWeaponRules(): { id: string; text: string; teams: string[]; byTeam: Record<string, string> }[] {
  return [...rareRegistry.entries()].map(([id, v]) => ({
    id,
    text: v.text,
    teams: [...v.teams],
    byTeam: Object.fromEntries(v.byTeam),
  }));
}

export function assertKnownRules(rules: WeaponRule[], where: string): void {
  for (const r of rules) {
    if (!isKnownWeaponRule(r.id)) {
      throw new Error(
        `Unknown weapon rule '${r.id}' (raw: '${r.raw}') on ${where}. ` +
          `Register it in the team module via registerRareWeaponRule() — weapon rules are never no-ops.`,
      );
    }
  }
}

/** Distance-prefixed Devastating, e.g. `1" Devastating 3`. */
const DEV_DIST = /^(\d+(?:\.\d+)?)"\s*Devastating\s*(\d+)$/i;
/** `Heavy (Dash only)` and friends. */
const HEAVY_ONLY = /^Heavy\s*\(([^)]+?)\s*only\)$/i;
const PIERCING_CRITS = /^Piercing\s+Crits\s*(\d+)?$/i;
const SEEK_LIGHT = /^Seek\s+Light$/i;
const NAME_X = /^([A-Za-z' ’-]+?)\s*(\d+(?:\.\d+)?)(\+|")?$/;

/**
 * Parse one rule token as printed on a datacard (e.g. `Lethal 5+`, `Blast 2"`, `Range 9"`).
 * Returns null for empty tokens; the caller decides whether that is an error.
 */
export function parseWeaponRule(rawInput: string): WeaponRule | null {
  const raw = rawInput.trim().replace(/’/g, "'").replace(/[”“]/g, '"');
  if (!raw) return null;

  let m = DEV_DIST.exec(raw);
  if (m) return { id: 'Devastating', x: Number(m[2]), dist: Number(m[1]), raw };

  m = HEAVY_ONLY.exec(raw);
  if (m) return { id: 'Heavy', only: m[1]!.trim(), raw };

  m = PIERCING_CRITS.exec(raw);
  if (m) return { id: 'PiercingCrits', x: m[1] ? Number(m[1]) : 1, raw };

  if (SEEK_LIGHT.test(raw)) return { id: 'SeekLight', raw };

  m = NAME_X.exec(raw);
  if (m) {
    const id = canonicalRuleId(m[1]!.trim());
    return { id, x: Number(m[2]), raw };
  }

  return { id: canonicalRuleId(raw), raw };
}

function canonicalRuleId(name: string): WeaponRuleId {
  const cleaned = name.replace(/\s+/g, ' ').trim();
  const pascal = cleaned
    .split(/[\s-]+/)
    .map((w) => w.charAt(0).toUpperCase() + w.slice(1).toLowerCase())
    .join('');
  // Known aliases seen on datacards.
  const aliases: Record<string, WeaponRuleId> = {
    PiercingCrits: 'PiercingCrits',
    SeekLight: 'SeekLight',
  };
  return (aliases[pascal] ?? pascal) as WeaponRuleId;
}

/** Parse a comma/semicolon separated rules cell. */
export function parseWeaponRules(text: string): WeaponRule[] {
  if (!text) return [];
  return text
    .split(/[,;]/)
    .map((s) => parseWeaponRule(s))
    .filter((r): r is WeaponRule => r !== null);
}

/** Human-readable rendering, round-trips the printed form. */
export function formatWeaponRule(r: WeaponRule): string {
  if (r.raw) return r.raw;
  if (r.id === 'Devastating' && r.dist !== undefined) return `${r.dist}" Devastating ${r.x}`;
  if (r.id === 'PiercingCrits') return `Piercing Crits ${r.x ?? 1}`;
  if (r.id === 'SeekLight') return 'Seek Light';
  if (r.id === 'Heavy' && r.only) return `Heavy (${r.only} only)`;
  if (r.id === 'Lethal') return `Lethal ${r.x}+`;
  if (r.id === 'Range' || r.id === 'Blast' || r.id === 'Torrent') return `${r.id} ${r.x}"`;
  return r.x !== undefined ? `${r.id} ${r.x}` : r.id;
}

/**
 * Condensed Environment (Close Quarters): "Weapons with the Blast, Torrent and/or
 * x" Devastating weapon rule also have the Lethal 5+ weapon rule."
 */
export function condensedEnvironmentRules(rules: WeaponRule[]): WeaponRule[] {
  const triggers = rules.some(
    (r) => r.id === 'Blast' || r.id === 'Torrent' || (r.id === 'Devastating' && r.dist !== undefined),
  );
  if (!triggers) return rules;
  if (rules.some((r) => r.id === 'Lethal' && (r.x ?? 6) <= 5)) return rules;
  return [...rules, { id: 'Lethal' as const, x: 5, raw: 'Lethal 5+ (Condensed Environment)' }];
}
